### PR TITLE
Adding support for per-task PodTemplate configuration

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
This PR adds support for configuring task executions using a per-task defined `PodTemplate` and / or `PodTemplate` name. The differences are client vs. server side application respectively. 

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Currently supported task types:
- [ ] container
- [ ] dask
- [x] mpi
- [x] pod
- [x] python
- [x] pytorch
- [ ] ray
- [x] shell
- [ ] spark
- [x] tensorflow

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3123

## Follow-up issue
_NA_

### CODE REVIEW

**Review by ByronHsu — COMMENTED** [review]
nit: the name of [this function](https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow_test.go#L73) is forgotten to change. Can you also fix this?

File: go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow.go [hunk 1]
```diff
@@ -90,15 +82,15 @@ func (tensorflowOperatorResourceHandler) BuildResource(ctx context.Context, task
...
 				RestartPolicy: commonOp.RestartPolicyNever,
 			},
 			kubeflowv1.TFJobReplicaTypeChief: {
 				Replicas: &chiefReplicas,
 				Template: v1.PodTemplateSpec{
-					ObjectMeta: objectMeta,
+					ObjectMeta: *objectMeta,
 					Spec:       *podSpec,
 				},
 				RestartPolicy: commonOp.RestartPolicyNever,
```
[the remark this answers is not in the document]
Comment (hamersaw) [reply go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow.go:90]: It should be, added it!

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go [hunk 2]
```diff
@@ -145,26 +146,75 @@ func ToK8sPodSpec(ctx context.Context, tCtx pluginsCore.TaskExecutionContext) (*
 	return pod, nil
 }
 
-func MergePodSpecs(podTemplatePodSpec *v1.PodSpec, podSpec *v1.PodSpec, primaryContainerName string) (*v1.PodSpec, error) {
-	if podTemplatePodSpec == nil || podSpec == nil {
-		return nil, errors.New("podTemplatePodSpec and podSpec cannot be nil")
+// getBasePodTemplate attempts to retrieve the PodTemplate to use as the base for k8s Pod configuration. This value can
+// come from on of the following:
```
[the remark this answers is not in the document]
Comment (hamersaw) [reply go/tasks/pluginmachinery/flytek8s/pod_helper.go]: thanks!

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go [hunk 3]
```diff
@@ -104,67 +107,230 @@ func UpdatePod(taskExecutionMetadata pluginsCore.TaskExecutionMetadata,
...
+		if primaryContainerName, ok = taskTemplate.GetConfig()[PrimaryContainerKey]; !ok {
+			return nil, nil, "", pluginserrors.Errorf(pluginserrors.BadTaskSpecification,
+				"invalid TaskSpecification, config missing [%s] key in [%v]", PrimaryContainerKey, taskTemplate.GetConfig())
+		}
+
+		// update annotations and labels
+		if taskTemplate.GetK8SPod().Metadata != nil {
+			objectMeta.Annotations = utils.UnionMaps(objectMeta.Annotations, taskTemplate.GetK8SPod().Metadata.Annotations)
+			objectMeta.Labels = utils.UnionMaps(objectMeta.Labels, taskTemplate.GetK8SPod().Metadata.Labels)
+		}
```
Comment (EngHabu) [inline go/tasks/pluginmachinery/flytek8s/pod_helper.go:163]: I'm probably missing something, we don't seem to assign any values to objectMeta before these lines, why do we need to union?

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go [hunk 4]
```diff
@@ -104,67 +107,230 @@ func UpdatePod(taskExecutionMetadata pluginsCore.TaskExecutionMetadata,
...
+	objectMeta := metav1.ObjectMeta{
+		Annotations: make(map[string]string),
+		Labels:      make(map[string]string),
 	}
-	if task.GetContainer() == nil {
-		logger.Errorf(ctx, "Default Pod creation logic works for default container in the task template only.")
-		return nil, fmt.Errorf("container not specified in task template")
+	primaryContainerName := ""
+
+	if taskTemplate.GetContainer() != nil {
```
Comment (EngHabu) [inline go/tasks/pluginmachinery/flytek8s/pod_helper.go:126]: Switch it to a `switch target := taskTemplate.GetTarget().(type)` block to avoid casting twice and maybe improve readability that it's a "oneof" ?

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go [hunk 5]
```diff
@@ -104,67 +107,230 @@ func UpdatePod(taskExecutionMetadata pluginsCore.TaskExecutionMetadata,
...
+			primaryContainer = &podSpec.Containers[index]
+		}
+	}
+
+	if primaryContainer == nil {
+		return nil, nil, pluginserrors.Errorf(pluginserrors.BadTaskSpecification, "invalid TaskSpecification, primary container [%s] not defined", primaryContainerName)
+	}
+
+	// add copilot configuration to primaryContainer and PodSpec (if necessary)
+	if taskTemplate.GetContainer() != nil {
```
Comment (EngHabu) [inline go/tasks/pluginmachinery/flytek8s/pod_helper.go:213]: I understand most of this code is just refactored to a new location but why is this only happening in case of a container not a Pod task?

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go (hunk 3 above)
Comment (hamersaw) [reply go/tasks/pluginmachinery/flytek8s/pod_helper.go]: I updated this to merge the `taskTemplate` annotations and labels into the `objectMeta`. Would you prefer just setting them explicitly?

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go (hunk 4 above)
Comment (hamersaw) [reply go/tasks/pluginmachinery/flytek8s/pod_helper.go]: Great idea, done!

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go (hunk 5 above)
Comment (hamersaw) [reply go/tasks/pluginmachinery/flytek8s/pod_helper.go:222]: Yeah, so the recent [flytekit PR](https://github.com/flyteorg/flytekit/pull/1425) does not yet address adding the `pod_template` support to `ContainerTasks`. The difficulty here is that the `DataLoadingConfig` struct is [only available on the Container message](https://github.com/flyteorg/flyteidl/blob/c60305cb0ced352cf3918f5e0e3de116271e0479/protos/flyteidl/core/tasks.proto#L197), so not on the K8sPod. This means that currently we can only support injecting copilot uploader / downloader for `Container` instances of the `TaskTemplate`.

The plan was to address this when we implement `pod_template` for the `ContainerTask`, there have been multiple community members willing to work on it immediately. This will require minor refactoring here - ie. add `DataLoadingConfig` to either `TaskTemplate` (and deprecate the field in `Container`) or `K8sPod` proto message, and update this code to use the correct value.

Alternatively, happy to keep this PR open until we add support for `pod_template` in the `ContainerTask` which will require the aforementioned updates.

**Review by EngHabu — APPROVED** [review]

**Comment by fg91** [comment]
Thanks a lot @hamersaw 🙏 🙂

**Comment by ByronHsu** [comment]
Thank you so much @hamersaw

**Comment by flixr** [comment]
very nice, thanks @hamersaw :smile:
